File: server/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "cmd")]
pub enum IncomingMessage {
    #[serde(rename = "login")]
    Login(LoginCommand),

    #[serde(rename = "sync")]
    Sync { device_pts: u64 },

    #[serde(rename = "send_message")]
    SendMessage(SendMessageCommand),
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LoginCommand {
    pub user_token: String,
    pub device_token: String,
    pub secret_key: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SendMessageCommand {
    pub receiver: String,
    pub contents: String,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "event")]
pub enum OutgoingMessage {
    #[serde(rename = "device_online")]
    DeviceOnline,

    #[serde(rename = "device_offline")]
    DeviceOffline,

    #[serde(rename = "sync_updates")]
    SyncUpdates(SyncUpdates),

    #[serde(rename = "update")]
    Update(Update),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct SyncUpdates {
    pub too_long: bool,
    pub synced: bool,
    pub updates: Vec<Update>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Update {
    pub pts: u64,
    pub payload: UpdatePayload,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum UpdatePayload {
    #[serde(rename = "new_message")]
    NewMessage(NewMessageUpdate),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct NewMessageUpdate {
    pub sender: String,
    pub contents: String,
}

File: mac-client/BeepPager/BeepPager/Protocol.swift
//
//  Created by Cyandev on 2024/9/29.
//  Copyright (c) 2024 Cyandev. All rights reserved.
//

import Foundation

fileprivate struct _StringKey: CodingKey {
    
    var stringValue: String
    var intValue: Int?
    
    init(stringValue: String) {
        self.stringValue = stringValue
    }
    
    init?(intValue: Int) {
        return nil
    }
}

enum OutgoingMessage: Codable {
    case login(LoginCommand)
    case sync(SyncCommand)
    
    func encode(to encoder: any Encoder) throws {
        var commandType: String
        switch self {
            case .login(let loginCommand):
                try loginCommand.encode(to: encoder)
                commandType = type(of: loginCommand).type
            case .sync(let syncCommand):
                try syncCommand.encode(to: encoder)
                commandType = type(of: syncCommand).type
        }
        
        var container = encoder.container(keyedBy: _StringKey.self)
        try container.encode(commandType, forKey: .init(stringValue: "cmd"))
    }
}

protocol Command {
    
    static var type: String { get }
}

struct LoginCommand: Codable, Command {
    
    static let type: String = "login"
    
    var userToken: String
    var deviceToken: String
    var secretKey: String
    
    enum CodingKeys: String, CodingKey {
        case userToken = "user_token"
        case deviceToken = "device_token"
        case secretKey = "secret_key"
    }
}

struct SyncCommand: Codable, Command {
    
    static let type: String = "sync"
    
    var devicePts: UInt64
    
    enum CodingKeys: String, CodingKey {
        case devicePts = "device_pts"
    }
}

enum IncomingMessage: Decodable {
    case loggedIn
    case syncUpdates(SyncUpdates)
    
    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: _StringKey.self)
        let eventType = try container.decode(String.self, forKey: .init(stringValue: "event"))
        
        struct _UnknownEventType: Error { }
        
        switch eventType {
        case "logged_in":
            self = Self.loggedIn
        case "sync_updates":
            self = .syncUpdates(try SyncUpdates(from: decoder))
        default:
            throw _UnknownEventType()
        }
    }
}

struct SyncUpdates: Codable {
    
    var tooLong: Bool
    var isSynced: Bool
    var updates: [Update]
    
    enum CodingKeys: String, CodingKey {
        case tooLong = "too_long"
        case isSynced = "synced"
        case updates = "updates"
    }
}

struct Update: Codable {
    
    // TODO: implement this.
}
